### PR TITLE
Update protocol doc for new VOPRF API (fixes #30)

### LINKED ISSUE #30: Update usage of VOPRF API
API has been updated at https://github.com/cfrg/draft-irtf-cfrg-voprf/ quite substantially.

### PR DESCRIPTION
**Pull request by alxdavids**
- Rewrite parts of intro
- Clean up protocol overview
- Rewrite API functions
- Rewrite protocol instantiation
- Update ciphersuites

(This is a big change, apologies!)

cc @dvorak42, @chris-wood for reviews (will also notify mailing list).

### CODE REVIEW

File: draft-davidson-pp-protocol.md [hunk 1]
```diff
@@ -815,223 +589,105 @@ and is necessary for achieving the unforgeability requirement.
...
-  `as_elements()` to individually deserialize each of the individual
-  byte arrays into a single group element and output a new array
-  containing each of these elements.
-
-## API instantiation {#voprf-api}
+## Recommended ciphersuites {#voprf-ciph-recs}
 
-For the explicit signatures of each of the functions, refer to
-{{pp-functions}}.
+We recommend that the Issuer chooses one of the following ciphersuites
```
Comment (chris-wood) [inline draft-davidson-pp-protocol.md:597]:
```suggestion
The RECOMMENDED Issuer ciphersuites are as follows:
```

Comment (chris-wood) [reply draft-davidson-pp-protocol.md:597]: Probably one of them should be MTI too, right?

File: draft-davidson-pp-protocol.md [hunk 2]
```diff
@@ -815,223 +589,105 @@ and is necessary for achieving the unforgeability requirement.
...
-                s_id: id
-                ciphersuite: ciph,
-                pub_key: pub_key,
-                max_evals: max_evals
-            }
-8. Output (s_cfg, s_update)
-~~~
+We deliberately avoid the usage of smaller ciphersuites (associated with
+P-256 and curve25519) due to the potential to reduce security via
+static-DH attacks. See {{I-D.irtf-cfrg-voprf}} for more details.
```
Comment (chris-wood) [inline draft-davidson-pp-protocol.md:606]:
```suggestion
static Diffie Hellman attacks. See {{I-D.irtf-cfrg-voprf}} for more details.
```

File: draft-davidson-pp-protocol.md [hunk 3]
```diff
@@ -815,223 +589,105 @@ and is necessary for achieving the unforgeability requirement.
...
-              },
-              ClientIssuanceElement {
-                  msg_data: i_data,
-              }
-          }
+def Generate(m):
+  inputs = []
+  for i in range(m):
+    inputs[i] = random_bytes()
+  (tokens, blindedTokens) = Blind(inputs)
```
Comment (chris-wood) [inline draft-davidson-pp-protocol.md:631]: As we removed the batch interface to Blind, this pseudocode probably needs to be updated.

File: draft-davidson-pp-protocol.md [hunk 4]
```diff
@@ -815,223 +589,105 @@ and is necessary for achieving the unforgeability requirement.
...
-8. N = ciph.VerifiableUnblind(G,pk,M,Z,r,proof)
-9. if N == "error": panic(ERR_PROOF_VALIDATION)
-10. tokens = []
-11. for i in 0..m:
-        1. issued = N[i].as_bytes()
-        2. rt = RedemptionToken { data: c_data[i], issued: issued }
-        3. tokens[i] = rt
-12. Output tokens
+Process(pkI, input, resp):
+  unblindedTokens = Unblind(pkI, input.data, input.msg, resp)
```
Comment (chris-wood) [inline draft-davidson-pp-protocol.md:653]: Ditto here for the batch interface.

File: draft-davidson-pp-protocol.md [hunk 5]
```diff
@@ -1047,32 +703,34 @@ on the underlying operations in a non-black-box manner. Hence, an
...
+The ciphersuites that we describe for the Privacy Pass protocol are
+derived from the core instantiations of the protocol (such as in
+{{voprf-protocol}}).
+
+In each of the ciphersuites below, the maximum security provided
+corresponds to the maximum difficulty of computing a discrete logarithm
+in the group. Note that the actual security level MAY be lower, see the
+security considerations in {{I-D.irtf-cfrg-voprf}} for examples.
+
+## PP(OPRF2)
```
Comment (chris-wood) [inline draft-davidson-pp-protocol.md:717]: Where did "OPRF2" come from?

File: draft-davidson-pp-protocol.md (hunk 1 above)
Comment (alxdavids) [reply draft-davidson-pp-protocol.md]: How is this usually specified?

File: draft-davidson-pp-protocol.md (hunk 4 above)
Comment (alxdavids) [reply draft-davidson-pp-protocol.md:653]: Right, but if we have a batched proof then we'll need to evaluate it over all the evaluated elements at once. I don't think this is will be as easy as pushing it all into the for loop?

File: draft-davidson-pp-protocol.md (hunk 5 above)
Comment (alxdavids) [reply draft-davidson-pp-protocol.md:717]: Due to OPRF ciphersuite with ID=0x0002. I could just call it by its full name, but I thought that it introduced naming redundancy.

**Pull request merged**